torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -1440,19 +1464,13 @@ Tensor sparse_sparse_matmul_backward(
   TORCH_CHECK(
       grad_order == 0 || grad_order == 1,
       ": grad_order not in [0, 1] at sparse_sparse_matmul_backward function");
-  const auto mask_ones_like = [](const Tensor& t) -> Tensor {
-    return at::sparse_coo_tensor(
-        t._indices(),
-        at::ones({1}, t._values().options()).expand_as(t._values()),
-        t.sizes());
-  };
 
   if (grad_order == 0) {
     auto a_grad = _sparse_sparse_matmul(grad, b.conj().t());
-    return a_grad.mul(mask_ones_like(a.coalesce()));
+    return sparse_mask_like_grad(a, a_grad);
```

Comment:
Ok, so essentially we're starting to write manual fusions. If we had torch.compile support we could presumably generate the code for this pattern. Since this is a hot path this is ok, but it's going to start to become difficult to test correctness.

Since `_sparse_mask_projection` is a native function you could even write a test that compares it to the pattern that you're fusing. 

Is this a common general pattern? Maybe we can apply it in more locations. If we generalize the fused composite native function that you're adding here a bit, maybe it applies to more locations? It can then also subsume all the logic of `sparse_mask_like_grad` and we can explicitly say "This function represents the fusion of this pattern". That'll be easier to understand to future maintainers.

Reply:
@cpuhrsch , alternatively, we can make this function composite implicit now that `sparse_mask` does support backward with COO inputs. This way we can remove the code for backward altogether. The result: much less code, still sync-less backward albeit potentially slower compared to this impl (there is only one way to call backward, there is no back-and-force between grad and input, because of the fixed semantics of `sparse_mask`). But, yes, it is a generic pattern forces by some "sparse-semantics" functions which might become obsolete when differentiable `sparse_mask`.
Another possibility: we can create a public function which is doing either `_sparse_mask_projection` and/or `sparse_mask` (whichever is faster) which is explicitly differentiable and could be used to enforce sparse semantics all while having flexible backward.

Reply:
I think we should go with your first approach for now and keep the amount of code low and then work on torch.compile support to provide these fusions. We can keep what you have for now as a reference and goal for that integration. Otherwise we'll have to undo these manual fusions again once we have torch.compile support.